en: add doc about volume snapshot backup/restore across multiple kubernetes (#2406)
pingcap/docs-tidb-operator: en/restore-by-ebs-snapshot-across-multiple-kubernetes.md

Reviewer: Preview: https://preview-operator-pingcap-doc.pingcap-docs-website-scaffold-1hq.pages.dev/tidb-in-kubernetes/dev/restore-by-ebs-snapshot-across-multiple-kubernetes

summary: Learn how to restore a TiDB cluster across multiple Kubernetes from EBS Volume Snapshots.
---

# Restore a TiDB Cluster across Multiple Kubernetes from EBS Volume Snapshots

This document describes how to restore backup data in AWS EBS snapshots to a TiDB cluster across multiple Kubernetes.

The restore method described in this document is implemented based on CustomResourceDefinition (CRD) in BR Federation Manager and TiDB Operator. For the underlying implementation, [BR](https://docs.pingcap.com/tidb/stable/backup-and-restore-overview) is used to restore the data. BR stands for Backup & Restore, which is a command-line tool for distributed backup and recovery of the TiDB cluster data.

## Limitations

- Snapshot restore is applicable to TiDB Operator v1.5.0 or later versions and TiDB v6.5.3 or later versions.
- You can use snapshot restore only to restore data to a cluster with the same number of TiKV nodes and volumes configuration. That is, the number of TiKV nodes and volume configurations of TiKV nodes are identical between the restore cluster and backup cluster.
- Snapshot restore is currently not supported for TiFlash, TiCDC, DM, and TiDB Binlog nodes.

> - Snapshot restore creates volumes with the default configuration (3000 IOPS/125 MB) of GP3. To perform restore using other configurations, you can specify the volume type or configuration, such as `--volume-type=gp3`, `--volume-iops=7000`, or `--volume-throughput=400`.

Author: 已经从 limitation 中去掉，并且加到 restore 过程中了 https://github.com/pingcap/docs-tidb-operator/pull/2406/commits/7eb44f2eaa3006d3f3e587eac50bd345c09ed07a#diff-f36f3bb1f0593b115746461afa83d36869730aef33bb88913f1d23645f69fc54R55

  ```yaml
  spec:
    template:
      br:
        sendCredToTikv: false
        options:
        - --volume-type=gp3
        - --volume-iops=7000
        - --volume-throughput=400
      serviceAccount: tidb-backup-manager
      toolImage: pingcap/br:v7.1.0
   ```

## Prerequisites

Before restoring a TiDB cluster across multiple Kubernetes from EBS volume snapshots, you need to complete the following preparations.

> ### Complete the volume backup
> 
> To restore a TiDB cluster across multiple Kubernetes from EBS snapshots, you should have a completed `VolumeBackup`. For steps of performing snapshot backup, refer to [Back Up a TiDB Cluster across Multiple Kubernetes Using EBS Volume Snapshots](backup-by-ebs-snapshot-across-multiple-kubernetes.md).
> 
> ### Prepare the restore clusters
> 
> Deploy a TiDB cluster across multiple Kubernetes to which you want to restore data. See [Deploy a TiDB Cluster across Multiple Kubernetes Clusters](deploy-tidb-cluster-across-multiple-kubernetes.md). **Note, You need add the `recoveryMode: true` field to spec of `TidbCluster`**.

Reviewer: ```suggestion
- Complete the volume backup

    For detailed steps, refer to [Back Up a TiDB Cluster across Multiple Kubernetes Using EBS Volume Snapshots](backup-by-ebs-snapshot-across-multiple-kubernetes.md).

- Prepare the restore cluster

    - Deploy a TiDB cluster across multiple Kubernetes that you want to restore data to. For detailed steps, refer to [Deploy a TiDB Cluster across Multiple Kubernetes Clusters](deploy-tidb-cluster-across-multiple-kubernetes.md).
    - When deploying the TiDB cluster, add the `recoveryMode: true` field to the spec of `TidbCluster`.
```

## Restore process

### Step 1. Set up the environment for EBS volume snapshot restore in every data plane

**You must execute the following steps in every data plane**.

> 1. Download the [backup-rbac.yaml](https://github.com/pingcap/tidb-operator/blob/master/manifests/backup/backup-rbac.yaml) file to the restore server.

Reviewer: 这里确认是 backup-rbac.yaml 不是 restore-rbac.yaml 吗？

Author: 是的，他们用相同的 rbac

> 2. Supposed that you deploy the TiDB cluster in `${namespace}`, create the RBAC-related resources required for the restore in this namespace by running the following command.

Reviewer: 这里的意思是需要把 RBAC 资源放在和 restore cluster 相同的 namespace 下吗？

Author: 是的，需要在相同 namespace 下

If you grant permissions by associating ServiceAccount with IAM, you can create the `VolumeRestore` CR as follows:

```shell
kubectl apply -f restore-fed.yaml
```

The `restore-fed.yaml` file has the following content:

```yaml
---
apiVersion: federation.pingcap.com/v1alpha1
kind: VolumeRestore
metadata:
  name: ${restore-name}
spec:
  clusters:
  - k8sClusterName: ${k8s-name1}
    tcName: ${tc-name1}
    tcNamespace: ${tc-namespace1}
    backup:
      s3:
        provider: aws
        region: ${region-name}
        bucket: ${bucket-name}
        prefix: ${backup-path1}
  - k8sClusterName: ${k8s-name2}
    tcName: ${tc-name2}
    tcNamespace: ${tc-namespace2}
    backup:
      s3:
        provider: aws
        region: ${region-name}
        bucket: ${bucket-name}
        prefix: ${backup-path2}
  - ... # other clusters
  template:
    br:
      sendCredToTikv: false
    toolImage: ${br-image}
    serviceAccount: tidb-backup-manager
```

</div>
</SimpleTab>

### Step 3. View the restore status

After creating the `VolumeRestore` CR, the restore process automatically start.

To check the restore status, use the following command:

```shell
kubectl get vrt -n ${namespace} -o wide

> ```

Reviewer: backup 的文档里有 **删除 CR** 的步骤，restore 是否要加上这一步？

Author: 我觉得 restore 不需要，backup 文档里介绍的原因是删除 CR 可以将备份数据一起删掉。

        - --volume-iops=7000
        - --volume-throughput=400
      serviceAccount: tidb-backup-manager
      toolImage: pingcap/br:v7.1.0
   ```

## Prerequisites

Before restoring a TiDB cluster across multiple Kubernetes from EBS volume snapshots, you need to complete the following preparations.

- Complete the volume backup

    For detailed steps, refer to [Back Up a TiDB Cluster across Multiple Kubernetes Using EBS Volume Snapshots](backup-by-ebs-snapshot-across-multiple-kubernetes.md).

- Prepare the restore cluster

    - Deploy a TiDB cluster across multiple Kubernetes that you want to restore data to. For detailed steps, refer to [Deploy a TiDB Cluster across Multiple Kubernetes Clusters](deploy-tidb-cluster-across-multiple-kubernetes.md).
    - When deploying the TiDB cluster, add the `recoveryMode: true` field to the spec of `TidbCluster`.

## Restore process

### Step 1. Set up the environment for EBS volume snapshot restore in every data plane

**You must execute the following steps in every data plane**.

1. Download the [backup-rbac.yaml](https://github.com/pingcap/tidb-operator/blob/master/manifests/backup/backup-rbac.yaml) file to the restore server.

> 2. Supposed that you deploy the TiDB cluster in `${namespace}`, create the RBAC-related resources required for the restore in this namespace by running the following command.

Reviewer: ```suggestion
2. Create the RBAC-related resources required for the restore by running the following command. Note that the RBAC-related resources must be put in the same `${namespace}` as the TiDB cluster. 
```
